src/ruis/render/scene/scene_renderer.cpp

```diff
@@ -0,0 +1,190 @@
+/*
+carcockpit - Car cockpit example GUI project
+
+Copyright (C) 2024 Gagistech Oy <gagisechoy@gmail.com>
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+/* ================ LICENSE END ================ */
+
+#include "scene_renderer.hxx"
+
+#include <chrono>
+
+#include "../../../carcockpit/application.hpp"
+
+using namespace ruis::render;
+
+scene_renderer::scene_renderer(utki::shared_ref<ruis::context> c) :
+	context_v(std::move(c))
+{
+	texture_default_black = context_v.get().loader.load<ruis::res::texture_2d>("texture_default_black").to_shared_ptr();
+	texture_default_white = context_v.get().loader.load<ruis::res::texture_2d>("texture_default_white").to_shared_ptr();
+	texture_default_normal =
+		context_v.get().loader.load<ruis::res::texture_2d>("texture_default_normal").to_shared_ptr();
+	texture_default_environment_cube =
+		context_v.get().loader.load<ruis::res::texture_cube>("tex_cube_env_hata").to_shared_ptr();
+
+	prepare_fullscreen_quad_vao();
+}
+
+void scene_renderer::set_scene_scaling_factor(ruis::real scene_scaling_factor)
+{
+	this->scene_scaling_factor = scene_scaling_factor;
+}
+
+void scene_renderer::set_scene(std::shared_ptr<ruis::render::scene> scene_v)
+{
+	this->scene_v = scene_v;
+}
+
+void scene_renderer::set_environment_cube(std::shared_ptr<ruis::res::texture_cube> texture_environment_cube)
+{
+	this->texture_environment_cube = texture_environment_cube;
+}
+
+void scene_renderer::set_external_camera(std::shared_ptr<ruis::render::camera> cam)
+{
+	external_camera = cam;
+}
+
+void scene_renderer::render(const ruis::rect& dims, const ruis::mat4& viewport_matrix)
```

Comment:
As far as I understand, rectangle is not needed here, it only needs dimensions, so pass `ruis::vec2` instead of `ruis::rect`.

Reply:
done